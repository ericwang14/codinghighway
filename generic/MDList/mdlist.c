#include <stdlib.h>
#include <stdio.h>

struct mdl_node {
  int val;
  struct mdl_node *child;
  struct mdl_node *next;
};

struct mdl_token {
  enum token_type {
    NUMBER,
    OPEN_P = '(',
    CLOSE_P = ')',
    PTR = '>',
    END
  } type;
  int n; /* meaningless if type != NUMBER */
} token;

void next_token(void) {
  int c;
  if ((c = getchar()) == EOF)
    token.type = END;
  if (c == '(')
    token.type = OPEN_P;
  else if (c == ')')
    token.type = CLOSE_P;
  else if (c == '>')
    token.type = PTR;
  else {
    ungetc(c, stdin);
    scanf("%d", &token.n);
    token.type = NUMBER;
  }
}

struct mdl_node *md_node(void);
struct mdl_node *md_list(void) {
  if (token.type == END)
    return NULL;
  struct mdl_node *n = md_node();
  if (token.type == PTR) {
    next_token();
    n->next = md_list();
  }
  return n;
}

struct mdl_node *md_node(void) {
  struct mdl_node *n = malloc(sizeof(*n));
  n->val = token.n;
  n->child = n->next = NULL;
  next_token();
  if (token.type == OPEN_P) {
    next_token();
    n->child = md_list();
    if (token.type != CLOSE_P)
      fprintf(stderr, "Missing close parentheses\n");
    else
      next_token();
  }
  return n;
}

void print_list(struct mdl_node *h) {
  while (h) {
    printf("%d", h->val);
    if (h->child) {
      printf("(");
      print_list(h->child);
      printf(")");
    }
    printf("->");
    h = h->next;
  }
  printf("END");
}

int main(void) {
  next_token();
  print_list(md_list());
  printf("\n");
  return 0;
}
