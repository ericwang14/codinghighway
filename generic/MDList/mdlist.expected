1
1>2
1>2
1>2>4>8>16>32>64>128>256>512>1024
1>2>4>8>16>32>64>128>256>512>1024
10>9>3>8>7>4>5>6>110>120>130>140>150>141
3>2>1
16>27>2>12>9>19>31>5>4>11>1>15>37>23>25>20>10>29>3>13>28>26>14>21>22>41>38>6>24>34>30>17>8>7>18>36>43>40>35>32>39>33>42
1>2>3>4>5>6>7>8>9>10>11>12>13>14>15>16>17>18>19>20>21>22>23>24>25>26>27>28>29>30>31>32>33
1>2>3>4>5>6>7>8>9>10>11>12>13>14>15>16>17>18>19>20
1>2>3>4>5>6>7>8>9>10>11>12>13>14>15>16>17>18>19>20
1>11>111>110>109>108>107>106>105>104>103>102>101>100
94>43>55>23>71>10>14>19>17>200>2001>2002>2003>2004>2005>-1>-2>-3>-4>-5>-6>-7>-8>-9>-10>-11>-12>-13>-14>-15>-16>-17>-18>-19>-20
94>43>55>23>71>10>14>19>17>200>2001>2002>2003>2004>2005>-1>-2>-3>-4>-5>-6>-7>-8>-9>-10>-11>-12>-13>-14>-15>-16>-17>-18>-19>-20
